scripts/embed_env_vars.py

```diff
@@ -0,0 +1,69 @@
+from utils import envutils
+
+Import('env')
+
+# This file is invoked by PlatformIO during build.
+# See 'extra_scripts' in 'platformio.ini'.
+
+# Get PIO variables.
+build_flags = env['BUILD_FLAGS'] or []
+project_dir = env['PROJECT_DIR']
+
+# By default, the build is run in DEVELOP mode.
+# If we are running in CI and either building the master branch,
+# or building a PR that will merge into the master branch, THEN
+# we will build in RELEASE mode.
+is_ci = envutils.get_bool('CI')
+is_branch_master = envutils.get_github_ref_name() == 'master'
+is_pr_into_master = envutils.get_github_base_ref() == 'master'
+is_release = is_ci and (is_branch_master or is_pr_into_master)
+
+# Get the build type string.
+build_type = 'release' if is_release else 'debug'
+
+dotenv = envutils.dotenv(project_dir, build_type)
+
+# Defines that will be passed to the compiler.
+cpp_defines: dict[str, str | int | bool] = {}
+
+# Gets the build flags from the PIO ini file, removes the -D prefix, splits on the first '=', and then converts to a dictionary.
+leftover_flags = []
+for flag in build_flags:
+    if flag.startswith('-D'):
+        flag = flag[2::]
+        if '=' in flag:
+            (k, v) = flag.split('=', 1)
+
+            cpp_defines[k] = v
+        else:
+            cpp_defines[flag] = True
+    else:
+        leftover_flags.append(flag)
+
+# Gets all the environment variables prefixed with 'OPENSHOCK_'.
+openshock_vars = dotenv.get_all_prefixed('OPENSHOCK_')
+cpp_defines.update(openshock_vars)
+
+# Gets the log level from environment variables.
+(loglevel_int, _) = dotenv.get_loglevel('DEBUG_LEVEL')
```

Comment:
I like that you _do_ call it `loglevel` here but the env var is still called `DEBUG_LEVEL` XD

Reply:
Changed this